substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/linux/LinuxSubstitutions.java

```diff
@@ -66,6 +72,46 @@ public static String mapLibraryName(String libname) {
     }
 }
 
+@Platforms(Platform.LINUX.class)
+@TargetClass(java.lang.Runtime.class)
+final class Target_java_lang_Runtime {
+
+    @Substitute
+    public int availableProcessors() {
+        if (SubstrateOptions.MultiThreaded.getValue()) {
+            LinuxSched.cpu_set_t cpuSet;
+            int size, res;
+            final int pid = Unistd.getpid();
+            // try easy path
+            cpuSet = StackValue.get(LinuxSched.cpu_set_t.class);
+            size = SizeOf.get(LinuxSched.cpu_set_t.class);
+            res = LinuxSched.sched_getaffinity(pid, size, cpuSet);
+            if (res == 0) return LinuxSched.CPU_COUNT_S(size, cpuSet);
+            // try with more CPUs in a loop
+            size = Integer.highestOneBit(size) << 1;
+            while (Integer.numberOfTrailingZeros(size) < 16) { // we have to give up at *some* point
+                assert Integer.bitCount(size) == 1;
```

Comment:
Why do you need this assert?  Line 91 sets `size` to have exactly 1 bit set, and line 102 shifts that bit but does not add any new bits.  And you give up if there are 16 trailing zeroes, so I don't think line 102 can shift the bit off into the sign bit, for reasonable values of `__CPU_SETSIZE` and `__NCPUBITS`.  And if the bit does get shifted into the sign bit, the assert is not going to help.  Maybe the while-loop wants to give up when `size` gets to be more than you are willing to `calloc`.  That way you can have a nicely-named `final int` as the limit instead of a `16`.